src/software/util/design_patterns/generic_factory.cpp

```diff
@@ -0,0 +1,72 @@
+#include "software/util/design_patterns/generic_factory.h"
+
+// A quality of life typedef to make things shorter and more readable
+template <class IndexType, class TypeToCreate>
+using GenericRegistry = std::unordered_map<IndexType, std::function<std::unique_ptr<TypeToCreate>()>>;
```

Comment:
This is a duplicate typdef, `GenericRegistry` is already defined in the header. As a side note, I thought templated functions all had to be implemented in the header? My understanding is that this works because of `always_link=true` in the build target? @garethellis0

Reply:
Removed the duplicate typedef.

Reply:
So the rule-of-thumb (and it might even be stricter then that, but I'm not totally confident) is that if you use any `T` in your implementations (here it could be `IndexType` or `TypeToCreate`) then that needs to live in the header. If this is compiling it should be fine.